xbmc/pvr/channels/PVRChannelProviders.cpp

```diff
@@ -0,0 +1,370 @@
+    {
+      return provider;
+    }
+  }
+
+  return std::shared_ptr<CPVRChannelProvider>();
+}
+
+void CPVRChannelProvidersContainer::InsertEntry(
+    const std::shared_ptr<CPVRChannelProvider>& newProvider, ProviderUpdateMode updateMode)
+{
+  bool found = false;
+  for (auto& provider : m_providers)
+  {
+    if (provider->GetClientId() == newProvider->GetClientId() &&
+        provider->GetUniqueId() == newProvider->GetUniqueId())
+    {
+      found = true;
+      provider->UpdateEntry(newProvider, updateMode);
+    }
+  }
+
+  if (!found)
+  {
+    m_providers.emplace_back(newProvider);
+  }
+}
+
+CPVRChannelProviders::CPVRChannelProviders()
+{
+}
+
+bool CPVRChannelProviders::Load()
+{
+  // unload previous providers
+  Unload();
+
+  // load providers from database
+  bool bReturn = LoadFromDatabase();
+
+  // update from clients
+  Update();
+
+  return bReturn;
+}
+
+void CPVRChannelProviders::Unload()
+{
+  // remove all tags
+  CSingleLock lock(m_critSection);
+  m_providers.clear();
+}
+
+bool CPVRChannelProviders::Update()
+{
+  {
+    CSingleLock lock(m_critSection);
+    if (m_bIsUpdating)
+      return false;
+    m_bIsUpdating = true;
+  }
+
+  // Default providers are the add-ons themselves
+  CPVRClientMap clients;
+  int numClients = CServiceBroker::GetPVRManager().Clients()->GetClients(clients);
+  CLog::LogFC(LOGDEBUG, LOGPVR, "Adding default providers, found %d PVR add-ons", numClients);
+  CPVRChannelProviders newAddonProviderList;
+  for (const auto& clientEntry : clients)
+  {
+    auto addonProvider =
+        std::make_shared<CPVRChannelProvider>(clientEntry.second, clientEntry.first);
+    newAddonProviderList.CheckAndAddEntry(addonProvider, ProviderUpdateMode::BY_CLIENT);
+  }
+  UpdateAddonEntries(newAddonProviderList);
+
+  // Client providers are retrieved from the clients
+  CLog::LogFC(LOGDEBUG, LOGPVR, "Updating providers");
+  CPVRChannelProvidersContainer newProviderList;
+  std::vector<int> failedClients;
+  CServiceBroker::GetPVRManager().Clients()->GetChannelProviders(&newProviderList, failedClients);
+  return UpdateClientEntries(newProviderList, failedClients);
+}
+
+bool CPVRChannelProviders::PersistUserChanges(
+    const std::vector<std::shared_ptr<CPVRChannelProvider>>& providers)
+{
+  for (auto provider : providers)
+  {
+    provider->Persist(true);
+
+    CLog::LogFC(LOGDEBUG, LOGPVR, "Updated provider {} on client {}", provider->GetUniqueId(),
+                provider->GetClientId());
+  }
+
+  return true;
+}
+
+bool CPVRChannelProviders::LoadFromDatabase()
+{
+  const std::shared_ptr<CPVRDatabase> database = CServiceBroker::GetPVRManager().GetTVDatabase();
+  if (database)
+  {
+    bool bChanged = false;
+
+    CPVRChannelProviders providers;
+    database->Get(providers);
+
+    for (auto& provider : providers.GetAll())
+    {
+      // Shared ptr returned, !! will return 1 for an object and 0 for nullptr
+      bChanged |= !!CheckAndAddEntry(provider, ProviderUpdateMode::BY_DATABASE);
```

Comment:
Such tricks always need an additional line of explanation. Good, you did that. But you could avoid this extra comment if you not try to be super-clever and just write something anybody understands without additional comment. ;-)

`bChanged |= (CheckAndAddEntry(provider, ProviderUpdateMode::BY_DATABASE) != nullptr);`

Reply:
if _anybody_ should understand it, needs to be like this :-)
`bChanged = bChanged || CheckAndAddEntry(provider, ProviderUpdateMode::BY_DATABASE) != nullptr;`

Reply:
I stole the trick from the Timers class. But I far prefer the easier to read version!

Reply:
And done.